AI Assistant Block

glendaviesnz opened the pull request:
## Proposed changes:

* Adds a new AI Assistant block as a proposed replacement for the AI Paragraph and AI Image blocks.

https://user-images.githubusercontent.com/3629020/236058189-0e42f307-4db1-40d1-9362-b36d1df9369e.mp4


## Jetpack product discussion
More background here pesk38-qp-p2

## Todo

- [ ] Split Edit component up and split out the image and text portions into separate components that are imported into `edit.js`
- [ ] Look at removing the tag and category lookups now that the user is supplying the prompts
- [ ] Add error handling
- [ ] Add tests
- [ ] Investigate use of streaming from Open AI to reduce perceived latency

## Does this pull request change what data or activity we track or use?

Yes. It fires  `jetpack_ai_gpt3_completion`, `jetpack_ai_dalle_generation` and `jetpack_ai_dalle_generation_upload` events as the previous blocks.

## Testing instructions:

**For Jetpack sites**

Follow instructions in pe4Cmx-1cK-p2 under **How to develop AI blocks for Jetpack and review ongoing Pull Requests**

**For WordPress.com**

- Sync PR to Sandbox with beta blocks setting enabled and add the AI Assistant block to a post and try various prompts for text and images

glendaviesnz committed:
Add accept button to convert content to standard blocks

glendaviesnz committed:
[not verified] add user prompt input

glendaviesnz committed:
Add AI Assistant beta block as potential replacement for Paragraph block

glendaviesnz committed:
[not verified] Revert "[not verified] add user prompt input"

This reverts commit 2f3ec5533143336c14fce3e578ffb8216b8fc86e.

glendaviesnz committed:
[not verified] Revert "Add accept button to convert content to standard blocks"

This reverts commit d6aa68dff30d82d8d7e633b2b44d1f6e7c36c809.

glendaviesnz committed:
Remove debugging

glendaviesnz committed:
[not verified] Tidy up layout

glendaviesnz committed:
Switch to using markdown and hide ai paragraph and ai image in the inserter

glendaviesnz committed:
Add options to generate based on title or preceding content

glendaviesnz committed:
Add toggle to switch to image generation

glendaviesnz committed:
[not verified] Bring in image functionality

glendaviesnz committed:
[not verified] Update prompt and hide image and paragraph blocks in inserter

glendaviesnz committed:
[not verified] Add some todo comments

oskosk committed:
Merge branch 'trunk' into update/ai-paragraph-to-create-standard-blocks

oskosk committed:
changelog

oskosk commented:
Pushed rebase in [ed4012b](https://github.com/Automattic/jetpack/pull/30444/commits/ed4012b928e7e1fcde4ccc5f96c0312d7f0a7cea) and changelog in [1d885e5](https://github.com/Automattic/jetpack/pull/30444/commits/1d885e5c2196973563cfbb1dd736f3e54aaedcea)

oskosk commented:
Moved it out of draft

oskosk committed:
Factor out ImageWithSelect

oskosk committed:
Factor out ShowLittleByLittle

oskosk committed:
Factor out getImagesFromOpenAI

mtias commented on projects/plugins/jetpack/extensions/blocks/ai-assistant/attributes.js:
```diff
@@ -0,0 +1,10 @@
+		type: 'string',
+		source: 'html',
+	},
+	animationDone: {
```
This is a bit weird as a block attribute instead of block state

oskosk replied:
Fixed in 39057169044e471dbdc795cf2c860222d6aea0b7

mtias commented on projects/plugins/jetpack/extensions/blocks/ai-assistant/index.js:
```diff
@@ -0,0 +1,67 @@
+		_x( 'GPT', 'block search term', 'jetpack' ),
+		_x( 'AL', 'block search term', 'jetpack' ),
+		_x( 'Magic', 'block search term', 'jetpack' ),
+		_x( 'images', 'block search term', 'jetpack' ),
```
let's skip these about images and add one for "help"

lhkowalski replied:
In the same line of "help", maybe "assistant" is useful as well. What do you think?

oskosk committed:
Fix some lint issues

oskosk committed:
Replace img for input image"

oskosk committed:
Make animationDone be a state variable instead of an attribute

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/ai-paragraph-to-create-standard-blocks

jeherve committed:
Avoid build error

jeherve committed:
Fix version

anomiex committed:
[not verified] Fix Jetpack stable version (#30514)

We forgot to bring https://plugins.trac.wordpress.org/changeset/2907132
back here previously.

retrofox commented:
I'm getting the following issue:

> "cURL error 28: Operation timed out after 10002 milliseconds with 0 bytes received"

I think it should be part of `Add error handling` or maybe we should re-attempt and/or increase the timeout?

retrofox commented on projects/plugins/jetpack/extensions/blocks/ai-assistant/edit.js:
```diff
@@ -0,0 +1,444 @@
+	type = 'userPrompt'
+) => {
+	const promptSuffix = __(
+		'. Please always output the generated content in markdown format. Do not include a top level heading by default. Please only output generated content ready for publishing.',
```
Is it ok defining the prompt here? I'm not saying it's a blocker but should we build it on the server side? 
Just FYI, this line in some way is triggering the failing test:

https://github.com/Automattic/jetpack/actions/runs/4917595635/jobs/8782896803?pr=30444#step:9:501

```
expect(received).toBeFalsy()
  [test-extensions] 
      Received: ". Please always output the generated content in markdown format. Do not include a top level heading by default. Please only output generated content ready for publishing."
```

oskosk replied:
It could be built server-side eventually indeed. There's [an issue](https://github.com/orgs/Automattic/projects/667/views/1?pane=issue&itemId=27456467) about it in [the board](https://github.com/orgs/Automattic/projects/667/views/1)

dhasilva replied:
Just a side note, I like how polite it is.

dhasilva replied:
Trying portuguese prompts here and the language mismatch between the prompt and the suffix is not a problem, so it does not need to be translated as well, another reason to do it server-side.

retrofox committed:
[not verified] create and export propmt suffix const

dhasilva reviewed (commented):
I'm getting a lot of timeout errors when testing locally (4~10), but otherwise it works :+1:

retrofox committed:
[not verified] iterate over createPrompt() function. Update tests.

retrofox committed:
[not verified] comment MAX content test

retrofox commented:
@glendaviesnz I've updated a little bit the `createPrompt()` function. Updated some tests and commented the other ones.
I'm not sure if they are properly defined or we should consider to re-write them. 
Please let me know how we can continue helping here.

oskosk commented on projects/plugins/jetpack/extensions/blocks/ai-assistant/edit.js:
```diff
@@ -50,6 +64,10 @@ export const createPrompt = (
+	if ( ! postTitle?.length ) {
+		return '';
+	}
+
```
The lack of title shouldn't be a blocker for generating the prompt. If there's a new post without a title set it should still work

glendaviesnz commented:
> @glendaviesnz I've updated a little bit the `createPrompt()` function. Updated some tests and commented the other ones. I'm not sure if they are properly defined or we should consider to re-write them. Please let me know how we can continue helping here.

@retrofox Thanks for checking in with me. I am happy for you all to take over this PR and do what you think is best. I was just trying to put together some ideas about how the AI blocks could move forward at a point when it didn't look like anyone was giving them some love. Now a team is on the job, no need to keep me in the loop - go for it 😃  But by all means ping me if anything I have done/suggested is confusing or broken and you can't work it out.

lhkowalski committed:
Update block keywords

renatoagds committed:
Jetpack AI: Comment unused variables
